### ISSUE TITLE
Slower 1x when training with one GPU using Pytorch-xla compared with using pytorch only

### ISSUE THREAD

**Issue by yxd886** [pytorch/xla#3266]
## 🐛 Bug

I am training a CNN model using Pytorch-xla, and found it is 1x slower when training with one GPU using Pytorch-xla compared with using pytorch only.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: Tesla V100 GPU
 - torch_xla version: 1.11
 - using gcr.io/tpu-pytorch/xla:nightly_3.7_cuda_10.2

The metric report is listed below:
Metric: CompileTime
  TotalSamples: 9
  Accumulator: 45s226ms933.445us
  ValueRate: 01s405ms211.447us / second
  Rate: 0.279638 / second
  Percentiles: 1%=677ms829.248us; 5%=677ms829.248us; 10%=677ms829.248us; 20%=799ms875.615us; 50%=01s182ms111.622us; 80%=12s677ms819.667us; 90%=15s160ms011.186us; 95%=15s160ms011.186us; 99%=15s160ms011.186us
Metric: DeviceLockWait
  TotalSamples: 42
  Accumulator: 01s233ms330.471us
  ValueRate: 024ms077.639us / second
  Rate: 0.819943 / second
  Percentiles: 1%=001.957us; 5%=002.064us; 10%=002.113us; 20%=002.333us; 50%=002.875us; 80%=004.050us; 90%=227ms956.023us; 95%=234ms396.071us; 99%=271ms757.485us
Metric: ExecuteTime
  TotalSamples: 42
  Accumulator: 03s059ms840.721us
  ValueRate: 085ms703.399us / second
  Rate: 1.16304 / second
  Percentiles: 1%=010ms051.967us; 5%=033ms755.363us; 10%=033ms830.563us; 20%=034ms440.405us; 50%=046ms636.368us; 80%=077ms359.905us; 90%=276ms993.158us; 95%=277ms483.498us; 99%=287ms556.266us
Metric: InboundData
  TotalSamples: 36
  Accumulator: 468.00MB
  ValueRate: 14.15MB / second
  Rate: 1.08818 / second
  Percentiles: 1%=1.00B; 5%=1.00B; 10%=1.00B; 20%=1.00B; 50%=1.00B; 80%=3.00MB; 90%=75.00MB; 95%=75.00MB; 99%=75.00MB
Metric: InputOutputAliasCount
  TotalSamples: 9
  Accumulator: 1222.00
  ValueRate: 33.07 / second
  Rate: 0.243524 / second
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=1.00; 80%=495.00; 90%=495.00; 95%=495.00; 99%=495.00
Metric: IrValueTensorToXlaData
  TotalSamples: 238
  Accumulator: 01s139ms734.413us
  ValueRate: 022ms888.564us / second
  Rate: 4.5748 / second
  Percentiles: 1%=247.552us; 5%=277.773us; 10%=291.386us; 20%=312.389us; 50%=410.090us; 80%=003ms010.679us; 90%=005ms282.331us; 95%=011ms763.502us; 99%=038ms343.206us
Metric: OutboundData
  TotalSamples: 307
  Accumulator: 760.79MB
  ValueRate: 14.47MB / second
  Rate: 5.8389 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=8.00B; 50%=1.00KB; 80%=2.25MB; 90%=3.00MB; 95%=9.00MB; 99%=75.00MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 122
  Accumulator: 323ms541.040us
  ValueRate: 006ms247.782us / second
  Rate: 2.3632 / second
  Percentiles: 1%=148.881us; 5%=164.072us; 10%=177.354us; 20%=201.861us; 50%=395.734us; 80%=584.634us; 90%=872.132us; 95%=033ms211.167us; 99%=036ms926.469us
Metric: TensorsGraphSize
  TotalSamples: 42
  Accumulator: 39689.00
  ValueRate: 1101.07 / second
  Rate: 1.16519 / second
  Percentiles: 1%=166.00; 5%=166.00; 10%=166.00; 20%=166.00; 50%=271.00; 80%=366.00; 90%=5492.00; 95%=5980.00; 99%=5980.00
Metric: TransferFromServerTime
  TotalSamples: 36
  Accumulator: 02s807ms943.335us
  ValueRate: 055ms618.916us / second
  Rate: 1.08818 / second
  Percentiles: 1%=267.245us; 5%=291.883us; 10%=299.973us; 20%=325.684us; 50%=498.723us; 80%=011ms576.903us; 90%=288ms788.291us; 95%=290ms680.348us; 99%=290ms854.833us
Metric: TransferToServerTime
  TotalSamples: 307
  Accumulator: 01s355ms106.981us
  ValueRate: 026ms040.100us / second
  Rate: 5.89939 / second
  Percentiles: 1%=244.008us; 5%=271.223us; 10%=283.649us; 20%=305.473us; 50%=370.170us; 80%=002ms005.689us; 90%=005ms335.035us; 95%=012ms542.289us; 99%=038ms329.067us
Metric: TransferToServerTransformTime
  TotalSamples: 307
  Accumulator: 254ms504.559us
  ValueRate: 005ms821.453us / second
  Rate: 5.83889 / second
  Percentiles: 1%=027.008us; 5%=030.628us; 10%=033.670us; 20%=039.053us; 50%=056.451us; 80%=407.713us; 90%=950.169us; 95%=002ms398.179us; 99%=017ms866.493us
Counter: CachedCompile
  Value: 33
Counter: CreateCompileHandles
  Value: 9
Counter: CreateDataHandles
  Value: 3044
Counter: CreateXlaTensor
  Value: 16678
Counter: DestroyDataHandles
  Value: 2508
Counter: DestroyXlaTensor
  Value: 16078
Counter: DeviceDataCacheMiss
  Value: 59
Counter: MarkStep
  Value: 6
Counter: ReleaseDataHandles
  Value: 2508
Counter: UncachedCompile
  Value: 9
Counter: XRTAllocateFromTensor_Empty
  Value: 45
Counter: XrtCompile_Empty
  Value: 128
Counter: XrtExecuteChained_Empty
  Value: 128
Counter: XrtExecute_Empty
  Value: 128
Counter: XrtMemoryInfo_Empty
  Value: 128
Counter: XrtRead_Empty
  Value: 128
Counter: XrtReleaseAllocationHandle_Empty
  Value: 128
Counter: XrtReleaseCompileHandle_Empty
  Value: 128
Counter: XrtSessionCount
  Value: 10
Counter: XrtSubTuple_Empty
  Value: 128
Counter: aten::_local_scalar_dense
  Value: 24
Counter: aten::im2col
  Value: 6
Counter: aten::im2col_backward
  Value: 6
Counter: xla::_copy_from
  Value: 5323
Counter: xla::_s_where
  Value: 18
Counter: xla::_softmax
  Value: 12
Counter: xla::_softmax_backward_data
  Value: 12
Counter: xla::_to_cpu
  Value: 36
Counter: xla::_unsafe_view
  Value: 24
Counter: xla::abs
  Value: 6
Counter: xla::add
  Value: 2559
Counter: xla::addcdiv_
  Value: 738
Counter: xla::addcmul
  Value: 738
Counter: xla::addmm
  Value: 6
Counter: xla::all
  Value: 24
Counter: xla::bmm
  Value: 72
Counter: xla::clamp
  Value: 84
Counter: xla::clone
  Value: 72
Counter: xla::constant_pad_nd
  Value: 138
Counter: xla::convolution_backward_overrideable
  Value: 360
Counter: xla::convolution_overrideable
  Value: 360
Counter: xla::div
  Value: 1044
Counter: xla::empty
  Value: 628
Counter: xla::eq
  Value: 24
Counter: xla::erf
  Value: 24
Counter: xla::exp
  Value: 30
Counter: xla::expand
  Value: 108
Counter: xla::fill_
  Value: 180
Counter: xla::ge
  Value: 84
Counter: xla::gt
  Value: 72
Counter: xla::leaky_relu
  Value: 18
Counter: xla::leaky_relu_backward
  Value: 18
Counter: xla::log
  Value: 12
Counter: xla::lt
  Value: 78
Counter: xla::mean
  Value: 24
Counter: xla::mm
  Value: 12
Counter: xla::mse_loss
  Value: 6
Counter: xla::mse_loss_backward
  Value: 6
Counter: xla::mul
  Value: 2964
Counter: xla::neg
  Value: 96
Counter: xla::norm
  Value: 744
Counter: xla::permute
  Value: 132
Counter: xla::pow
  Value: 252
Counter: xla::relu
  Value: 96
Counter: xla::round
  Value: 12
Counter: xla::sgn
  Value: 6
Counter: xla::sigmoid
  Value: 12
Counter: xla::sigmoid_backward
  Value: 12
Counter: xla::slice
  Value: 48
Counter: xla::sqrt
  Value: 774
Counter: xla::stack
  Value: 6
Counter: xla::sub
  Value: 162
Counter: xla::sum
  Value: 30
Counter: xla::t
  Value: 24
Counter: xla::threshold_backward
  Value: 96
Counter: xla::transpose
  Value: 48
Counter: xla::uniform_
  Value: 102
Counter: xla::unsqueeze
  Value: 72
Counter: xla::view
  Value: 414
Counter: xla::zero_
  Value: 891
Metric: XrtAllocateFromTensor
  TotalSamples: 307
  Accumulator: 703ms792.391us
  Mean: 002ms289.226us
  StdDev: 030ms431.018us
  Rate: 5.89936 / second
  Percentiles: 25%=038.619us; 50%=051.755us; 80%=307.123us; 90%=721.516us; 95%=001ms420.221us; 99%=018ms894.063us
Metric: XrtCompile
  TotalSamples: 9
  Accumulator: 45s109ms337.013us
  Mean: 05s012ms148.557us
  StdDev: 05s384ms580.768us
  Rate: 0.279627 / second
  Percentiles: 25%=920ms119.686us; 50%=01s179ms139.570us; 80%=12s640ms510.452us; 90%=15s144ms481.070us; 95%=15s144ms481.070us; 99%=15s144ms481.070us
Metric: XrtExecute
  TotalSamples: 42
  Accumulator: 03s006ms330.881us
  Mean: 072ms579.307us
  StdDev: 078ms795.614us
  Rate: 1.16303 / second
  Percentiles: 25%=033ms832.048us; 50%=045ms278.137us; 80%=077ms942.652us; 90%=275ms180.945us; 95%=277ms692.500us; 99%=285ms495.325us
Metric: XrtReadLiteral
  TotalSamples: 36
  Accumulator: 01s036ms288.443us
  Mean: 029ms785.790us
  StdDev: 062ms728.976us
  Rate: 1.09214 / second
  Percentiles: 25%=071.573us; 50%=088.148us; 80%=006ms270.324us; 90%=167ms047.652us; 95%=169ms978.134us; 99%=169ms111.791us
Metric: XrtReleaseAllocation
  TotalSamples: 122
  Accumulator: 011ms949.342us
  Mean: 089.749us
  StdDev: 142.557us
  Rate: 2.36216 / second
  Percentiles: 25%=009.062us; 50%=019.694us; 80%=143.700us; 90%=263.158us; 95%=422.155us; 99%=633.652us

I am suspecting it is because of these three 
Counter: aten::_local_scalar_dense
  Value: 24
Counter: aten::im2col
  Value: 6
Counter: aten::im2col_backward
  Value: 6

I am curious why aten::im2col aten::im2col_backward are invoked?

**Comment by JackCaoG** [comment]
Yea, we have https://github.com/pytorch/xla/issues/2932 to track the lowering of `im2col` and backward. I think it is from `F.unfold`

**Comment by yxd886** [comment]
Dear JackCaoG,
    Thanks for your quick reply, yes, I found F.unfold in my code, and I also found that torch.distributions.normal.Normal.cdf torch.distributions.normal.Normal.pdf in my code will directly cause aten::_local_scalar_dense, is there any way to avoid this?

Thanks very much.

**Comment by JackCaoG** [comment]
I don't have a good solution for `unfold`, `_local_scalar_dense` happens when you convert a tensor to the scalar value (for example trying to print the value of a tensor). It is actually not the end of the world if it is not happening in the middle of the training.

If you can share the a sample code of `torch.distributions.normal.Normal.pdf` will trigger the `_local_scalar_dense` I can take a look tmr

**Comment by byronyi** [comment]
I plan to work on supporting im2col with torch-xla.
